### ISSUE TITLE
onemkl linear solver doesn't have a default difference quotient jacobian?

### ISSUE THREAD

**Issue by ciropom** [LLNL/sundials#182]
Hello again, 
I finally was able to compile link and execute my model with the onemkl dense linear solver.
The problem now is that at initialization time I get this error
```
[CVLS ERROR]  cvLsInitialize
  No Jacobian constructor available for SUNMatrix type


[CVODE ERROR]  cvInitialSetup
  The linear solver's init routine failed.
```

I'm using the following initialization code:

```
    /* Create dense SUNMatrix for use in linear solves */
    A = SUNMatrix_OneMklDense(y_len, y_len, SUNMEMTYPE_DEVICE, memhelper, &s->qspccQueue, s->sunctx);
	if (check_flag((void *)A, "SUNMatrix_OneMklDense", 0)) return s;
    s->matrix = A; /* save for deallocation */
	// Create the SUNLinearSolver object for use by CVode
    LS = SUNLinSol_OneMklDense(s->y, A, s->sunctx);
    if (check_flag((void *)LS, "SUNLinSol_OneMklDense", 0)) return s;
	s->flag = CVodeSetLinearSolver(s->cvode_mem, LS, A);
    if(check_flag(&s->flag, "CVodeSetLinearSolver", 1)) return s;    
```

It looks like onemkldense solver doesn't have a default difference-quotient jacobian, like the standard dense/band/lapack solvers.
Is this true? Users are forced to provide a Jacobian through CVodeSetJacFn to use this solver?

**Comment by gardner48** [comment]
Yes, that's correct. A user-supplied Jacobian function is currently required when using the oneMKL interface. This function may compute the analytic Jacobian or an approximation to it e.g., using difference-quotients.

**Comment by ciropom** [comment]
Why there is not any pre-packaged difference-quotient jacobian as it is for the standard dense solver?

**Comment by balos1** [comment]
> Why there is not any pre-packaged difference-quotient jacobian as it is for the standard dense solver?

The implementation of the difference-quotients routine in CVODE currently is specific to the standard dense and banded matrices. Generalizing it to support other matrices is something that I think would be nice, but we don't currently have any plans to do so.

**Comment by ciropom** [comment]
Thank you for the explanation. This is blocking in my scenario, where I'm translating automatically from MATLAB to C and simulating with sundials by using [QSPcc](https://github.com/cosbi-research/QSPcc)

Can you give me some hints on how to implement a difference-quotient jacobian for the MKL matrix?
Thanks

**Comment by drreynolds** [comment]
@ciropom: you will need to emulate the code from `cvLsDenseDQJac` ([here](https://github.com/LLNL/sundials/blob/main/src/cvode/cvode_ls.c#L1034)), but ensuring that each of the operations use device-resident vector data, and that these operations fill device-resident data within the MKL matrix.

**Comment by t0masoni** [comment]
Hello again, I don't know if something changed here?

**Comment by drreynolds** [comment]
@t0masoni, can you elaborate on your last comment?  As per the previous correspondence, SUNDIALS does not provide a built-in difference quotient approximation for onemkl matrices, but there is nothing to keep users from implementing this themselves.  Were you asking whether we had added one of these over the intervening time period?

**Comment by t0masoni** [comment]
Hello, I'm sorry that I was not clear. I was asking if the discussion is still true, or in the meantime something changed with respect to MKL interoperability.

**Comment by drreynolds** [comment]
Thanks for clarifying @t0masoni.  While we've kept up the current onemkl interface, we have not expanded it, so in that sense the status is unchanged.  At the present time, nobody on our team is working on research projects that rely on this interface; if a future project arises that uses this then we may develop upgrades.
